WangscapeTest/TestCartesianPower.cpp

```diff
@@ -0,0 +1,262 @@
+#include <gtest/gtest.h>
+
+#include <vector>
+
+#include <tilegen/CartesianPower.h>
+
+class TestCartesianPower : public ::testing::Test
+{
+protected:
+    typedef std::vector<int> Container;
+    typedef std::vector<int>::const_iterator Iterator;
+
+    Container base{4, 5, 6};
+    const tilegen::CartesianPower<Iterator> cp{base, 4};
+    TestCartesianPower()
+    {
+
+    };
+    ~TestCartesianPower() {};
+};
+
+TEST_F(TestCartesianPower, TestCartesianPowerSize)
+{
+    EXPECT_EQ(81, cp.size()) <<
+        "CartesianProduct has incorrect size";
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerCount)
+{
+    size_t count = 0;
+    for (const auto& it : cp)
+        count++;
+    EXPECT_EQ(81, count);
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerComparisonBeginEnd)
+{
+    EXPECT_EQ(cp.cbegin(), cp.cbegin());
+    EXPECT_NE(cp.cend(), cp.cbegin());
+    EXPECT_NE(cp.cbegin(), cp.cend());
+    EXPECT_EQ(cp.cend(), cp.cend());
```

Comment:
It's still too vague for me.There is the same object twice here, `cp`, and `cend()` is called two times, creating two iterators from the same set of arguments. They're compared with `operator==`, which internally check equality of them basing on each object's atrributes. They should have the same values (as both iterators were created from same arguments). Why should we check if `cp.cend() == cp.cend()`?